### ISSUE TITLE
Scenario explorer - sort / filter options

### ISSUE THREAD

**Issue by BenRQ** [NPBruce/valkyrie#963]
This issue should be used to discuss the sort and filters to implement in the new "scenario explorer" screen.
Root task is #962

**Description:**

In order to facilitate scenario selection for players, we will implement a sort option and filter options in the scenario explorer.

**Sort :** 
This is the list of options I have in mind, feel free to share if you think one is missing, or if some are useless :  
 - sort by rating
 - sort by duration (or average duration ?)
 - sort by difficulty (or average win ratio ?)
 
In term of presentation, I don't know yet how we can do it.

**Filter:**
Purpose of filter is to filter scenarios based on a criteria.
When clicking on a filter, it will select only scenario with this filter.
It will then be possible to add more option. The more filter you select, the less scenario you will have listed.

Right new, I'm thinking of two possible filter types
  - each languages via a flag
  - each expansions via a symbol (like the base game)

In term of presentation, I think we can have a button filter, that will display the filter options.
You then select your choice, validate, and see the list of scenario as a result.
Selecting the filter in the main screen would be better, but I don't see how we can fit everything without taking too much space.

Any ideas, proposal is welcome :)

**Comment by redwolf2** [comment]
The problem with filtering the downloaded quests is, that the process still relies on a lot of http calls for the meta files to be made. It would take some time to get to the result: "No results found".

> In term of presentation, I think we can have a button filter, that will display the filter options.
You then select your choice, validate, and see the list of scenario as a result.
Selecting the filter in the main screen would be better, but I don't see how we can fit everything without taking too much space.

Button filter are a nice idea. Check out the "chip buttons" on google image search.

**Comment by scrubbless** [comment]
Sort/filter by selected content
Sort/filter by  X expansion

**Comment by BenRQ** [comment]
@scrubbless, some questions:
what do you mean by "selected content"? Why sort and filter ?

Expansion is already in my list for filter, but I don't think it makes sense to sort it by expansion.

**Comment by BenRQ** [comment]
Feedback from @antontimmermans on this:

In one of the corners you can put an 'settings' icon. That icon gives a pop-up menu that lets you choose the following:

    Scenario languages: with tickboxes for each language that the player can sellect; Multiple boxes can be selected. The effect of selecting a tick box(es) is that a scenario containing the language(s) will appear in the list.
    Sorting: Choose in which way the list is sorted. Select only one option out of:
        A-Z;
        Difficulty (easy to difficult);
        Duration (short to long).
        User rating
        In case of sorting ends up with same primary sorting keys in the list:
            Difficulty level sorting; then the second key to sort on would be duration, third key: A-Z
            Duration level sorting; then the second key to sort on would be difficulty, third key: A-Z
            A-Z sorting; then the second key to sort on would be difficulty,third key: duration (<- default sorting)
            Note: that I didn't include the reverse selection (z-a; long to short, .. etc). I personally think that would be overkill and too much detail.
            Note: An additional switch could be added to selectwriters rating/user rating: witch results in:
            scenario duration vs. average duration
            scenario difficulty v.s. average win

Further setting that affect the scenario overview list is the available expansions which are set elsewhere in the main menu. You can choose to apply this selection as an extra filter on the list to only display the scenarios that MUST have a specific expansion and scenarios that 'add variation'.

**Comment by BenRQ** [comment]
I've been working on this, and I'm slowly getting there.

Sort and filter are now working.
Works online and offline.

Supported :  
 filter by languages and expansions
 sort by name, difficulty, duration, rating

Texts are not final of course, so feel free to comment.
Once decided, I will then need some translations.

Please note I've not update the scenario presentation yet, this will come last.

See some screenshots below : 

Default option : by rating, descending
![image](https://user-images.githubusercontent.com/40794717/50490807-65c97700-0a0f-11e9-8c68-36193c62386f.png)

Filter options:
![image](https://user-images.githubusercontent.com/40794717/50490839-8f829e00-0a0f-11e9-8d5e-85621e9219d0.png)

Sort options online:
![image](https://user-images.githubusercontent.com/40794717/50490875-b5a83e00-0a0f-11e9-984f-e815753466a2.png)


When offline:
![image](https://user-images.githubusercontent.com/40794717/50490748-197e3700-0a0f-11e9-8965-d3462bcef54e.png)

**Comment by antontimmermans** [comment]
@BenRQ looking good! The filter screen and the sort and filter buttons look very nice.

On the sort page: place Avg Duration below Duration and Update date below Name to keep matching criteria close to eachother. 
The second row of boxes would then be: Ratting | Update date | Virctory ratio | Avg Duration

In your example I miss the "update date" so I am not sure what that looks like. I am more in favour to have a marker/icon/button  that shows if an update is available or not. But I guess you might be solving this in the new scenario selection layout (#961)

I like the default sorting setting.

**Comment by BenRQ** [comment]
Yes, good idea for the change of names.

The "update date" is something new.
I'd like to be able to sort scenarios by the latest update date, to know what are the latest scenarios, what scenarios have been updated, and maybe play / replay them.
On the technical side, I still don't know how I will get the date.
I may need to add it during package generation as I didn't find how I could get the "modified date" from github website.

**Comment by redwolf2** [comment]
Adding a new element for it inside quest.ini is a good idea. The file modification date may be unreliant.

**Comment by antontimmermans** [comment]
To see if an update is available can also be visible on the scenario selection screen as proposed in #961 
That would work better for me than a sorting option, If it where up to me you could leave out the sort by "update date" option.

**Question**: if you sorted by a primary key selection and you have two items with same value (for example rating) what is then your secondary sorting key (and third).?

**Also for consideration**: To give new scenarios a fair chance I would advice to show the rating only after it is played ~10 times. 
Reason for this is that, if the first person rating it scores extreme, he/she influences the likelyhood of others playing it. For example: if the reating is extreme low ( maybe just because he/she lost the game or didn't pay much attention to clues) then not many will start the scenario and the score won't change much. 
After about 10 games the scores likely give a more fair rating. Below 10 times played you can list only the number of times played.

The scenario writer can still look in the statistics document and read the score/comments and update the  scenario during this 10-games-grace period.

**Comment by redwolf2** [comment]
The algorithm should also punish scenarios, that set a "modify date" higher than the current date, by moving them at the end of the list.

**Comment by antontimmermans** [comment]
@redwolf2 This just sounds to me like another another reason not to sort/filter by update date; I would go for an indicator in the scenario explorer that an update is available (see also #961)

**Comment by BenRQ** [comment]
I would like to have this date generated rather than enter manually.

I see 3 solutions : 
1) find a way to get file date from github through HTTP (I did not find a way to do this)
2) add a date in the ini file when the package is created this would be quite reliable I think, but some authors may abuse the system (I don't think they will to be honest, but I can add a security)
3) Manually add a creation date for each scenario in the manifest.ini file managed by @scrubbless , with the date of creation. This would not allow us to follow upgrade version though.

@antontimmermans , the purpose is not to "inform" user about an update, as they will get a button for this, if they have already downloaded the scenario.
I see 2 "use cases":
 - a player who want to see if a new scenario is available
 - a player who has already played the best rated scenario and is looking for a new one: an average note with a recent upgrade, you might want to play it to give it a chance.

Regarding your remark on "10 notes" before showing the rating, I agree though that we should give new scenario a fair chance.
10 rating is still quite a lot, it could take months before scenarios reach that number of ratings.
See the stats today, there are still 6 scenarios with less than 10 notes.
I would say something between 3 and 5 is the right number.

**Comment by BenRQ** [comment]
I found a way to get the latest commit date in github, so now this filter is also working ! 
Seeing the results, I'm still thinking it's interesting to have this filter.

Filter by date descending:
![image](https://user-images.githubusercontent.com/40794717/50543153-6c065180-0bcf-11e9-80ab-8e1d6da8bd92.png)

I've also implemented all filters related to statistics and improve the layout a little bit.
And added an info for the number of scenario filtered out.

![image](https://user-images.githubusercontent.com/40794717/50543162-c43d5380-0bcf-11e9-9316-45b421457eb9.png)


![image](https://user-images.githubusercontent.com/40794717/50543149-3e210d00-0bcf-11e9-9387-b4ea9eca14c5.png)


Only remaining filter I need to implement is 'author'.
This has been demanded by some, so I will do it but it will require the generation of a new package for each scenario to have the author in the ini file.
I will only take the first line of quest.authors to use for the comparison.

**Comment by BenRQ** [comment]
Sort by authors is now supported.
I tried to use the "quest.authors" data of the existing quests for this, but it was not possible.
Too much variety in the author description format depending on quests.

So I've created a new input field "quest.short_authors" for this.
This should be one line of text, that will be used for scenarios.
If not available, a text "Unknown authors" will be displayed.

If you have a better proposal for the input field name, and description, I'm happy to change it.

![image](https://user-images.githubusercontent.com/40794717/50574386-2375b080-0de8-11e9-996e-c1fb0afea4a2.png)


And I'm happy to say that filters and sort options are now fully functional :)

Translations still needs to be done for the following texts (I will take care of French)  : 

```
SORT_TITLE,Sort
SORT_SELECT_CRITERIA,Select the sort criteria:
SORT_SELECT_ORDER,Order by:
SORT_ASCENDING,Ascending
SORT_DESCENDING,Descending

SORT_BY_AUTHOR,Author
SORT_BY_NAME,Name
SORT_BY_DIFFICULTY,Difficulty
SORT_BY_DURATION,Duration

SORT_BY_RATING,Rating
SORT_BY_AVERAGE_DURATION,Avg duration
SORT_BY_WIN_RATIO,Victory ratio
SORT_BY_DATE,Update date

FILTER_TITLE,Select filters
FILTER_SELECT_LANG,Select the language(s) you are speaking:
FILTER_MISSING_EXPANSIONS_ON,Hide scenarios with missing expansions  ☑
FILTER_MISSING_EXPANSIONS_OFF,Hide scenarios with missing expansions  ☐
FILTER_TEXT_NUMBER_OF_FILTERED_SCENARIO,{0} scenario(s) filtered out

AUTHORS_SHORT,Single text line for Authors (for "sort by author" option)
AUTHORS_UNKNOWN,Unknown authors (needs a scenario update by author)
```

**Comment by BenRQ** [comment]
Here is a technical version if some of you want to try out the sort/filter options and/or try out translations.
https://we.tl/t-loNuNTi2v7

If you want to propose a translation, text size is quite limited so it's better to try it out.

Texts files to modify can be found in valkyrie_Data\StreamingAssets\text
You can use the Localization.English.txt file to get the text to translate.

Please note that download mechanism is not implemented yet in this version, so you can get unexpected behaviour when selecting scenarios not downloaded.

Downloaded scenarios should be able to start.

**Comment by redwolf2** [comment]
GERMAN translation:

```
SORT_TITLE,Sortieren
SORT_SELECT_CRITERIA,Sortierkriterien auswählen:
SORT_SELECT_ORDER,Sortieren nach:
SORT_ASCENDING,Aufsteigend
SORT_DESCENDING,Absteigend

SORT_BY_AUTHOR,Autor
SORT_BY_NAME,Name
SORT_BY_DIFFICULTY,Schwierigkeit
SORT_BY_DURATION,Dauer

SORT_BY_RATING,Bewertung
SORT_BY_AVERAGE_DURATION,Ø Dauer
SORT_BY_WIN_RATIO,Siegrate
SORT_BY_DATE,Aktualisierungsdatum

FILTER_TITLE,Filter auswählen
FILTER_SELECT_LANG,Bitte Sprachen wählen:
FILTER_MISSING_EXPANSIONS_ON,Scenarien mit fehlenden Erweiterungen verstecken  ☑
FILTER_MISSING_EXPANSIONS_OFF,Scenarien mit fehlenden Erweiterungen verstecken  ☐
FILTER_TEXT_NUMBER_OF_FILTERED_SCENARIO,{0} Scenarien gefunden

AUTHORS_SHORT,Einzelne Textzeile für "Autoren" (für die "Sortieren nach Autor"-Option)
AUTHORS_UNKNOWN,Unbekannte Autoren (benötigt ein Update durch den Autor um dies zu hinterlegen)
```

**Comment by antontimmermans** [comment]
checked and saw the following behavior:
```
                   ascending             descending
name               nr>a>Z                Z>a>nr
difficultly        low>high              high>low
duration           low>high              high>low   (<- is this (low duration+high duration)/2?)
rating             none>low>high         *high>low>none*
update date        ??? cant check        ??? cant check
victory ratio      none>low>high         high>low>none

```
It didn't save my preferences but went back to default sorting (marked * in above table).

**Comment by antontimmermans** [comment]
checked offline behavior: 

When going offline when game is active--> online statistics remained available for sorting

When starting app when offline: thex offline appears, and the rating, update date, victory ratio, avg duration are red and cannot be selected.

FIltering seems to work properly on languages. 

Could not test the hide extentionsoption (I have them all)

**Comment by antontimmermans** [comment]
overall conclusion: The parts I could test seemed to behave as expected. I was a bit disappointed it didn't save my personal preference but went back to default sorting.

**Comment by BenRQ** [comment]
@redwolf2, I've tested the German translation and two texts are too long:
Do you think you can propose a shorter version ?

For the button I could check the text size and adapt the font size if you don't have an easy solution.

![image](https://user-images.githubusercontent.com/40794717/50613578-3d91ba80-0ede-11e9-9848-60899969fdb6.png)

![image](https://user-images.githubusercontent.com/40794717/50613620-5bf7b600-0ede-11e9-938c-cd879daeaea4.png)

**Comment by redwolf2** [comment]
Aktualisierungsdatum -> `Updatedatum`

Too long description -> `Textzeile "Autoren" (Relevant für die "Sortieren Nach"-Option):`

**Comment by BenRQ** [comment]
First text is now perfect 👍 

Second text is still too long : 

![image](https://user-images.githubusercontent.com/40794717/50618506-8607a300-0ef3-11e9-98a7-638a646804f5.png)

We may need to find another way to describe this.

**Comment by redwolf2** [comment]
`Textzeile "Autoren" (Für "Sortieren Nach"-Option):`

**Comment by antontimmermans** [comment]
What about naming the first (original) authors field: "Credits" since it is used to name authors and possibly testers, reviewers or translators.

And name the new author field for sorting simply "Author(s)"?

**Comment by antontimmermans** [comment]
> Thank you @redwolf2 for German translations.
> 
> Thank you for your feedback @antontimmermans, very useful as usual.
> 
> * Authors and update date are working, but with no visual feedback it's difficult to confirm if it's working properly I agree.
> * Saving preferences is not implemented, I was wondering if it was really useful, so I understand it is :)
>   It's not a big development I will do it
> * I did not think about this use case : starting online and losing connection later.
>   I don't really know if I should change current implementation though.
>   Did you think there was something unexpected in the current behaviour ?

Last bullet: I had no expected screen behavior in mind when testing the last bullet. I was merely testing for 'robustness': The application should not crash when choosing an 'online sorting'-option after the application lost its online connection. (<- note: I will also check this when you implement the download buttons :-) )
I am guessing you keep the information in memory on start-up, so it will continue to work (not crash) and no changes are needed. In my opinion this behavior is preferred.
If you reload the information when switching screens, then you should probably look into this behavior and change to the correct on/off-line representation when loading the scenario selection screen.
If you load the sorting information when sorting options are clicked, then you should probably even update the screen status upon detection of connection-loss.

**Comment by BenRQ** [comment]
This feature is now fully implemented and merged in 2.2. Test version coming soon.
Please open a new ticket if you find an issue in the test version provided.

@antontimmermans saving preference is now working.
